File: problems/dfs/on_tree/max_depth_binary_tree/attempt.py
class Node:
	def __init__(self, val, left = None, right = None):
		self.val = val;
		self.left = left;
		self.right = right;

def tree_max_depth(root):
	def dfs(root):
		if root is None:
			return 0
		return max(dfs(root.left), dfs(root.right)) + 1
	return dfs(root) - 1 if root else 0;

def build_tree(nodes, f):
    val = next(nodes)
    if val == "x":
        return None
    left = build_tree(nodes, f)
    right = build_tree(nodes, f)
    return Node(f(val), left, right)

if __name__ == "__main__":
    root = build_tree(iter(input().split()), int)
    res = tree_max_depth(root)
    print(res)

File: problems/dfs/on_tree/max_depth_binary_tree/attempt.js
"use strict";

class Node {
    constructor(val, left = null, right = null) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

function dfs(root) {
	if (root === null) {
		return 0;
	}
	return Math.max(dfs(root.left), dfs(root.right)) + 1;
}

function treeMaxDepth(root) {
    // WRITE YOUR BRILLIANT CODE HERE
    return root !== null ? dfs(root) - 1 : 0;
}

// this function builds a tree from input; you don't have to modify it
// learn more about how trees are encoded in https://algo.monster/problems/serializing_tree
function buildTree(nodes, f) {
	const val = nodes.next().value;
	if (val === "x") return null; 
	const left = buildTree(nodes, f);
	const right = buildTree(nodes, f);
	return new Node(f(val), left, right);
}

function splitWords(s) {
	return s === "" ? [] : s.split(" ");
}

function* main() {
    const root = buildTree(splitWords(yield)[Symbol.iterator](), parseInt);
	const res = treeMaxDepth(root);
    console.log(res);
}

class EOFError extends Error {}
{
    const gen = main();
    const next = (line) => gen.next(line).done && process.exit();
    let buf = "";
    next();
    process.stdin.setEncoding("utf8");
    process.stdin.on("data", (data) => {
        const lines = (buf + data).split("\n");
        buf = lines.pop();
        lines.forEach(next);
    });
    process.stdin.on("end", () => {
        buf && next(buf);
        gen.throw(new EOFError());
    });
}

